insonmnia/worker/metrics/metrics.go

```diff
@@ -0,0 +1,194 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/noxiouz/zapctx/ctxlog"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/sonm-io/core/insonmnia/hardware/disk"
+	"github.com/sonm-io/core/insonmnia/hardware/ram"
+	"github.com/sonm-io/core/insonmnia/worker/gpu"
+	"github.com/sonm-io/core/proto"
+	"github.com/sonm-io/core/util"
+	"github.com/sonm-io/core/util/multierror"
+	"go.uber.org/zap"
+)
+
+type Handler struct {
+	logger *zap.Logger
+
+	GPUs map[sonm.GPUVendorType]gpu.MetricsHandler
+	// in future, special handlers for other hardware units can be controlled by this plugin
+
+	mu        sync.Mutex
+	lastState *sonm.WorkerMetricsResponse
+}
+
+func NewHandler(ctx context.Context, GPUConfig map[string]map[string]string) (*Handler, error) {
+	handler := &Handler{
+		GPUs:   make(map[sonm.GPUVendorType]gpu.MetricsHandler),
+		logger: ctxlog.GetLogger(ctx).Named("metrix"),
+	}
+
+	handler.logger.Debug("initializing metrics handler")
+
+	for vendor := range GPUConfig {
+		typeID, err := gpu.GetVendorByName(vendor)
+		if err != nil {
+			return nil, err
+		}
+
+		h, err := gpu.NewMetricsHandler(ctx, typeID)
+		if err != nil {
+			handler.logger.Error("failed to initialize GPU metrics plugin", zap.String("vendor", vendor), zap.Error(err))
+			return nil, err
+		}
+
+		handler.logger.Debug("successfully created GPU handler", zap.String("vendor", vendor))
+		handler.GPUs[typeID] = h
+	}
+
+	return handler, nil
+}
+
+func (m *Handler) Run(ctx context.Context) {
+	go func() {
+		m.logger.Debug("starting metrics collection")
+
+		tk := util.NewImmediateTicker(time.Minute)
+		defer tk.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				m.logger.Warn("context cancelled", zap.Error(ctx.Err()))
+				return
+			case <-tk.C:
+				m.update(ctx)
+			}
+		}
+	}()
+}
+
+func (m *Handler) update(ctx context.Context) {
+	m.logger.Debug("updating hardware metrics")
+	isError := false
+
+	g, err := m.updateGPUMetrics()
+	if err != nil {
+		m.logger.Warn("failed to update GPU metrics", zap.Error(err))
+		isError = true
+	}
+
+	c, err := m.updateCPUMetrics()
+	if err != nil {
+		m.logger.Warn("failed to update CPU metrics", zap.Error(err))
+		isError = true
+	}
+
+	d, err := m.updateDiskMetrics(ctx)
+	if err != nil {
+		m.logger.Warn("failed to update disk metrics", zap.Error(err))
+		isError = true
+	}
+
+	r, err := m.updateRAMMetrics()
+	if err != nil {
+		m.logger.Warn("failed to update RAM metrics", zap.Error(err))
+		isError = true
+	}
+
+	// do not update metrics if some part is failed
```

Comment:
it's quite strange decision - we update separate state, but we do not update the whole cached state